### ISSUE TITLE
Recursive function definition raises error

### ISSUE THREAD

**Issue by sebastien** [doug-moen/curv#53]
It seems that curv currently does not allow recursive function definition (see [group thread](https://groups.google.com/forum/#!topic/curv/N99KRDfcmNU)). The following code

```
rand1 =  match [
    (x)   ->  frac(sin(x)*100000.0);
    (x,y) -> rand1(rand1(x) + rand1(y));
];
```

raises the following error:

```
ERROR: illegal recursive reference
at file "letrec.curv":
3|     (x,y) -> rand1(rand1(x) + rand1(y));
                ^^^^^          
```

**Comment by doug-moen** [comment]
Curv only supports recursive function definitions, not recursive data definitions. The 'illegal recursive reference' error is reported early in compilation: it happens while resolving names, before all identifiers have been looked up.

A definition like this causes an error, because it looks like a recursive data definition to the compiler:
```
rand1 =  f [
    (x)   ->  frac(sin(x)*100000.0);
    (x,y) -> rand1(rand1(x) + rand1(y));
];
```
If the compiler somehow knew that `f` is a function that returns another function during this phase of compilation, then it could treat this as a recursive function definition and generate the appropriate code. But it is not smart enough to do that. And `match` is just an ordinary function, it is not a keyword or special syntax, so the same problem happens if you replace `f` with `match`.

Until this is fixed, the workaround is to write:
```
rand1 x = x >> match [
    (x)   ->  frac(sin(x)*100000.0);
    (x,y) -> rand1(rand1(x) + rand1(y));
];
```

**Comment by sebastien** [comment]
Yes, that's also what Prolog or Erlang do, and it will be consistent with the let..in pattern assignment (ie    let (x,y)=(1,0)).
